src/caffe/test/test_gradient_based_solver.cpp

```diff
@@ -483,6 +495,29 @@ TYPED_TEST(AdaGradSolverTest, TestLeastSquaresUpdateWithEverythingAccum) {
 }
 
 template <typename TypeParam>
+class AdamSolverTest : public GradientBasedSolverTest<TypeParam> {
+  typedef typename TypeParam::Dtype Dtype;
+
+ protected:
+  virtual void InitSolver(const SolverParameter& param) {
+    this->solver_.reset(new AdamSolver<Dtype>(param));
+  }
+  virtual SolverParameter_SolverType solver_type() {
+    return SolverParameter_SolverType_ADAM;
+  }
+};
+
+TYPED_TEST_CASE(AdamSolverTest, TestDtypesAndDevices);
+TYPED_TEST(AdamSolverTest, TestAdamLeastSquares) {
+  typedef typename TypeParam::Dtype Dtype;
+  const Dtype kLearningRate = 1.0;
+  const Dtype kMomentum = 0.9;
+  // momentum2 is already set to 0.999
+  const Dtype kWeightDecay = 0.0;
+  this->TestLeastSquaresUpdate(kLearningRate, kWeightDecay, kMomentum);
+}
+
```

Comment:
Note: #2836 and #2887 modifies solver as well as introduces new share param and snapshot tests. To be consistent, let's also add 4 new test cases:

    TestLeastSquaresUpdateWithEverythingShare
    TestLeastSquaresUpdateWithEverythingAccumShare
    TestSnapshot
    TestSnapshotShare

For TestSnapshot, you can take a look at TestSnapshot in SGDSolverTest as an example. For the other 3 shared cases, you just need to add `this->share_ = true;` to the corresponding test cases.